File: examples/42-embedded-lua-scripting/cpp/src/main.cpp
// pmon — chapter 42: an embedded-Lua restart policy via sol2.
//
// The restart/backoff decision lives entirely in policy.lua, not in this
// binary: the host loads it into a *sandboxed* Lua 5.4 state (base/string/
// table/math only — no os, io, package, debug; load/loadfile/dofile/require/
// collectgarbage/loadstring removed even though their libraries were never
// opened, since a global with that name could otherwise be reintroduced by a
// careless policy), calls `on_exit(info)` for every simulated child exit, and
// prints exactly what the policy hands back. A real SIGHUP — blocked with
// sigprocmask and consumed as data through a signalfd + poll, the same
// pattern ch12/ch13 use for SIGCHLD — reloads the policy file with no
// restart of pmon itself: the running supervisor keeps its child-tracking
// state, only the decision logic changes underneath it.

#include <sol/sol.hpp>

#include <cerrno>
#include <chrono>
#include <cstdio>
#include <expected>
#include <format>
#include <optional>
#include <print>
#include <string>
#include <string_view>
#include <system_error>
#include <vector>

#include <poll.h>
#include <signal.h>
#include <sys/signalfd.h>
#include <unistd.h>

namespace {

using std::chrono::steady_clock;

[[nodiscard]] std::error_code last_error() {
    return {errno, std::system_category()};
}

// Every stdout write is a bare fflush away from a race with the parent shell
// that greps this process's redirected output file for a sentinel line (see
// demo.sh's `run()`) — std::println is fully buffered once stdout is not a
// tty, so every print here is followed by an explicit flush.
void say(std::string_view line) {
    std::println("{}", line);
    std::fflush(stdout);
}

[[noreturn]] void die(const std::string& msg) {
    std::println(stderr, "pmon: {}", msg);
    std::exit(1);
}

// ---------------------------------------------------------------------------
// Sandbox: open only base/string/table/math, then strip the globals that
// would let a policy touch the filesystem, spawn processes, or load new code
// even though their owning libraries (os, io, package, debug) were never
// opened in the first place.
// ---------------------------------------------------------------------------

sol::state make_sandbox() {
    sol::state lua;
    lua.open_libraries(sol::lib::base, sol::lib::string, sol::lib::table, sol::lib::math);
    lua["load"] = sol::lua_nil;
    lua["loadfile"] = sol::lua_nil;
    lua["dofile"] = sol::lua_nil;
    lua["require"] = sol::lua_nil;
    lua["collectgarbage"] = sol::lua_nil;
    lua["loadstring"] = sol::lua_nil;

    sol::table host = lua.create_named_table("host");
    host.set_function("log", [](const std::string& msg) { say("pmon: policy: " + msg); });
    host.set_function("now_ms", []() -> long long {
        return std::chrono::duration_cast<std::chrono::milliseconds>(
                   steady_clock::now().time_since_epoch())
            .count();
    });
    return lua;
}

// ---------------------------------------------------------------------------
// Errors, three ways (C++): fallible operations return std::expected<T,
// std::string>, converting a failed sol::protected_function_result into its
// sol::error message; die() is reserved for CLI-usage/OS-resource failures
// that have no data-driven recovery.
// ---------------------------------------------------------------------------

// Load `path` under a protected parse+run, then call on_load() if the policy
// defines one (also protected). Any failure here is a policy-load error.
[[nodiscard]] std::expected<void, std::string> load_policy(sol::state& lua, const std::string& path) {
    sol::protected_function_result loaded = lua.safe_script_file(path, sol::script_pass_on_error);
    if (!loaded.valid()) {
        sol::error err = loaded;
        return std::unexpected(err.what());
    }
    sol::protected_function on_load = lua["on_load"];
    if (on_load.valid()) {
        sol::protected_function_result ran = on_load();
        if (!ran.valid()) {
            sol::error err = ran;
            return std::unexpected(err.what());
        }
    }
    return {};
}

// One simulated child-exit event to feed on_exit().
struct Event {
    std::string_view name;
    long long exit_code;
    long long restarts;
    long long consecutive_failures;
    long long uptime_ms;
    long long last_backoff_ms;
};

struct Decision {
    std::string action;
    long long delay_ms = 0;
    std::string reason;
};

// Build `info`, call the policy's on_exit(info) under a protected call, and
// read back action/delay_ms/reason. `signal` stays absent from the table
// (nil) for every event this chapter replays.
[[nodiscard]] std::expected<Decision, std::string> decide(sol::state& lua, const Event& ev) {
    sol::table info = lua.create_table();
    info["name"] = std::string{ev.name};
    info["exit_code"] = ev.exit_code;
    info["restarts"] = ev.restarts;
    info["consecutive_failures"] = ev.consecutive_failures;
    info["uptime_ms"] = ev.uptime_ms;
    info["last_backoff_ms"] = ev.last_backoff_ms;

    sol::protected_function fn = lua["on_exit"];
    if (!fn.valid()) {
        return std::unexpected(std::string{"policy defines no on_exit"});
    }
    sol::protected_function_result result = fn(info);
    if (!result.valid()) {
        sol::error err = result;
        return std::unexpected(err.what());
    }
    sol::table decision = result;
    return Decision{
        .action = decision["action"],
        .delay_ms = decision["delay_ms"],
        .reason = decision["reason"],
    };
}

void print_decision(std::string_view name, const Decision& dec) {
    say(std::format(R"(pmon: decision child={} action={} delay_ms={} reason="{}")", name, dec.action,
                     dec.delay_ms, dec.reason));
}

// ---------------------------------------------------------------------------
// SIGHUP, the systems-programming content: block it process-wide up front so
// there is never a window where it could interrupt a syscall or fire a
// default handler, then consume it as ordinary readability on a signalfd —
// same pattern ch12/ch13 use for SIGCHLD, just a different signal.
// ---------------------------------------------------------------------------

[[nodiscard]] std::expected<int, std::error_code> make_sighup_fd() {
    sigset_t mask;
    sigemptyset(&mask);
    sigaddset(&mask, SIGHUP);
    if (::sigprocmask(SIG_BLOCK, &mask, nullptr) != 0) {
        return std::unexpected(last_error());
    }
    const int fd = ::signalfd(-1, &mask, SFD_CLOEXEC);
    if (fd < 0) {
        return std::unexpected(last_error());
    }
    return fd;
}

// Block until SIGHUP arrives on `sigfd`, draining exactly one signalfd_siginfo.
void wait_for_sighup(int sigfd) {
    pollfd pfd{.fd = sigfd, .events = POLLIN, .revents = 0};
    for (;;) {
        const int n = ::poll(&pfd, 1, -1);
        if (n < 0) {
            if (errno == EINTR) {
                continue;
            }
            die("poll: " + last_error().message());
        }
        signalfd_siginfo info{};
        if (::read(sigfd, &info, sizeof info) == sizeof info && info.ssi_signo == SIGHUP) {
            return;
        }
    }
}

// Fresh sandbox + host table + protected load + on_load(), all rebuilt from
// scratch — the simplest correct way to guarantee reload leaves no state
// from the previous policy behind.
[[nodiscard]] std::expected<sol::state, std::string> reload(const std::string& policy_path) {
    sol::state lua = make_sandbox();
    if (auto loaded = load_policy(lua, policy_path); !loaded) {
        return std::unexpected(loaded.error());
    }
    return lua;
}

// ---------------------------------------------------------------------------
// Subcommands
// ---------------------------------------------------------------------------

int run_phase(sol::state& lua, const std::vector<Event>& events) {
    for (const auto& ev : events) {
        auto dec = decide(lua, ev);
        if (!dec) {
            std::println(stderr, "pmon: policy error: {}", dec.error());
            return 1;
        }
        print_decision(ev.name, *dec);
    }
    return 0;
}

int cmd_run(const std::string& policy_path) {
    auto sigfd = make_sighup_fd();
    if (!sigfd) {
        std::println(stderr, "pmon: signalfd: {}", sigfd.error().message());
        return 1;
    }

    auto phase1 = reload(policy_path);
    if (!phase1) {
        std::println(stderr, "pmon: policy error: {}", phase1.error());
        return 1;
    }
    sol::state lua = std::move(*phase1);

    static const std::vector<Event> kPhase1{
        {.name = "web", .exit_code = 0, .restarts = 0, .consecutive_failures = 0, .uptime_ms = 120000, .last_backoff_ms = 0},
        {.name = "worker", .exit_code = 1, .restarts = 0, .consecutive_failures = 0, .uptime_ms = 9000, .last_backoff_ms = 0},
        {.name = "worker", .exit_code = 1, .restarts = 1, .consecutive_failures = 1, .uptime_ms = 9000, .last_backoff_ms = 500},
    };
    if (run_phase(lua, kPhase1) != 0) {
        return 1;
    }

    say("pmon: awaiting SIGHUP to reload policy");
    wait_for_sighup(*sigfd);
    say("pmon: reload requested");
    ::close(*sigfd);

    // Re-read the same --policy path — its content is now policy-v2.lua.
    auto phase2 = reload(policy_path);
    if (!phase2) {
        std::println(stderr, "pmon: policy error: {}", phase2.error());
        return 1;
    }
    lua = std::move(*phase2);

    static const std::vector<Event> kPhase2{
        {.name = "api", .exit_code = 1, .restarts = 0, .consecutive_failures = 0, .uptime_ms = 5000, .last_backoff_ms = 0},
        {.name = "worker", .exit_code = 1, .restarts = 2, .consecutive_failures = 2, .uptime_ms = 9000, .last_backoff_ms = 1000},
    };
    if (run_phase(lua, kPhase2) != 0) {
        return 1;
    }
    return 0;
}

int cmd_check(const std::string& policy_path) {
    sol::state lua = make_sandbox();
    if (auto loaded = load_policy(lua, policy_path); !loaded) {
        std::println("pmon: policy error: {}", loaded.error());
        std::fflush(stdout);
        return 1;
    }

    const Event probe{.name = "probe", .exit_code = 1, .restarts = 0, .consecutive_failures = 0,
                       .uptime_ms = 9000, .last_backoff_ms = 0};
    auto dec = decide(lua, probe);
    if (!dec) {
        std::println("pmon: policy error: {}", dec.error());
        std::fflush(stdout);
        return 1;
    }

    std::string version = lua["POLICY_VERSION"];
    say(std::format("pmon: policy ok version={}", version));
    return 0;
}

void usage() {
    std::println(stderr, "usage: app <run|check> --policy PATH");
}

} // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        usage();
        return 2;
    }
    const std::string_view subcommand = argv[1];
    if (subcommand != "run" && subcommand != "check") {
        usage();
        return 2;
    }

    std::optional<std::string> policy;
    for (int i = 2; i < argc; ++i) {
        const std::string_view arg = argv[i];
        if (arg == "--policy" && i + 1 < argc) {
            policy = argv[++i];
        } else {
            usage();
            return 2;
        }
    }
    if (!policy) {
        usage();
        return 2;
    }

    return subcommand == "run" ? cmd_run(*policy) : cmd_check(*policy);
}

File: examples/42-embedded-lua-scripting/rust/src/main.rs
// pmon -- a restart/backoff supervisor whose policy lives in Lua, not in this
// binary. `policy.lua` (and its SIGHUP-swapped sibling `policy-v2.lua`) is
// loaded into a *sandboxed* mlua state: only base/string/table/math are
// opened, and the loader globals that could read files or eval new code from
// inside the sandbox (`load`, `loadfile`, `dofile`, `require`,
// `collectgarbage`, `loadstring`) are removed after the fact. `os`, `io`,
// `package` and `debug` are simply never opened, so a policy that reaches for
// `os.execute` hits a nil global -- see `run_check` and `policy-evil.lua`.
//
// Errors, three ways: CLI parse failures `bail!` out of `parse_args`;
// fallible host-side operations return `anyhow::Result` composed with
// `.context()`; and every boundary crossing into Lua (`exec`, `.call`, table
// `get`) is an `mlua::Result` that a protected load/call turns into data
// instead of a panic, so a bad policy degrades to a reported error, not a
// crash of the supervisor.

use std::env;
use std::fs;
use std::process::exit;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::thread;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use anyhow::{Context, Result, bail};
use mlua::{Function, Lua, LuaOptions, StdLib, Table, Value};

/// Wall-clock milliseconds since the UNIX epoch, exposed to policies as
/// `host.now_ms()`. "Monotonic-ish" is all the contract asks for -- policies
/// only use it for human-readable timestamps, never for scheduling.
fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}

/// Build a fresh, sandboxed Lua state: only base/string/table/math opened
/// (`os`/`io`/`package`/`debug` are never opened at all), then strip the
/// loader globals that could otherwise read files or evaluate fresh source
/// from inside the sandbox even without a filesystem library.
fn build_sandbox() -> mlua::Result<Lua> {
    let lua = Lua::new_with(
        StdLib::STRING | StdLib::TABLE | StdLib::MATH,
        LuaOptions::default(),
    )?;

    for g in [
        "load",
        "loadfile",
        "dofile",
        "require",
        "collectgarbage",
        "loadstring",
    ] {
        lua.globals().set(g, Value::Nil)?;
    }

    let host = lua.create_table()?;
    host.set(
        "log",
        lua.create_function(|_, msg: String| {
            println!("pmon: policy: {msg}");
            Ok(())
        })?,
    )?;
    host.set("now_ms", lua.create_function(|_, ()| Ok(now_ms()))?)?;
    lua.globals().set("host", host)?;

    Ok(lua)
}

/// Load and execute the policy chunk under the sandbox (a protected load, so
/// a syntax error comes back as `Err` instead of aborting), then call its
/// optional `on_load()` hook. Both steps are `mlua::Result` all the way
/// through, so callers decide how to surface a failure.
fn load_policy(lua: &Lua, path: &str) -> mlua::Result<()> {
    let src = fs::read_to_string(path)
        .map_err(|e| mlua::Error::RuntimeError(format!("reading {path}: {e}")))?;
    lua.load(&src).set_name(path).exec()?;

    if let Ok(on_load) = lua.globals().get::<Function>("on_load") {
        on_load.call::<()>(())?;
    }
    Ok(())
}

/// One supervised-child exit event, exactly as the shared scenario spec
/// describes it. `signal` is never set in this chapter's scenario (every
/// event here is an exit-code death), so the `info` table simply never gets
/// a `signal` field -- matching "a nil field is absent from the table".
struct ExitEvent {
    name: &'static str,
    exit_code: i64,
    restarts: i64,
    consecutive_failures: i64,
    uptime_ms: i64,
    last_backoff_ms: i64,
}

fn info_table(lua: &Lua, ev: &ExitEvent) -> mlua::Result<Table> {
    let info = lua.create_table()?;
    info.set("name", ev.name)?;
    info.set("exit_code", ev.exit_code)?;
    info.set("restarts", ev.restarts)?;
    info.set("consecutive_failures", ev.consecutive_failures)?;
    info.set("uptime_ms", ev.uptime_ms)?;
    info.set("last_backoff_ms", ev.last_backoff_ms)?;
    Ok(info)
}

/// Call `on_exit(info)` under a protected call and print the resulting
/// decision. The `reason` text is entirely Lua's (built with
/// `string.format` in policy.lua) -- this only formats the wrapping
/// `pmon: decision ...` line around whatever the policy returned.
fn decide(lua: &Lua, ev: &ExitEvent) -> mlua::Result<()> {
    let info = info_table(lua, ev)?;
    let on_exit: Function = lua.globals().get("on_exit")?;
    let decision: Table = on_exit.call(info)?;

    let action: String = decision.get("action")?;
    let delay_ms: i64 = decision.get("delay_ms")?;
    let reason: String = decision.get("reason")?;

    println!(
        "pmon: decision child={} action={} delay_ms={} reason=\"{}\"",
        ev.name, action, delay_ms, reason
    );
    Ok(())
}

fn phase1(lua: &Lua) -> mlua::Result<()> {
    decide(
        lua,
        &ExitEvent {
            name: "web",
            exit_code: 0,
            restarts: 0,
            consecutive_failures: 0,
            uptime_ms: 120_000,
            last_backoff_ms: 0,
        },
    )?;
    decide(
        lua,
        &ExitEvent {
            name: "worker",
            exit_code: 1,
            restarts: 0,
            consecutive_failures: 0,
            uptime_ms: 9_000,
            last_backoff_ms: 0,
        },
    )?;
    decide(
        lua,
        &ExitEvent {
            name: "worker",
            exit_code: 1,
            restarts: 1,
            consecutive_failures: 1,
            uptime_ms: 9_000,
            last_backoff_ms: 500,
        },
    )?;
    Ok(())
}

fn phase2(lua: &Lua) -> mlua::Result<()> {
    decide(
        lua,
        &ExitEvent {
            name: "api",
            exit_code: 1,
            restarts: 0,
            consecutive_failures: 0,
            uptime_ms: 5_000,
            last_backoff_ms: 0,
        },
    )?;
    decide(
        lua,
        &ExitEvent {
            name: "worker",
            exit_code: 1,
            restarts: 2,
            consecutive_failures: 2,
            uptime_ms: 9_000,
            last_backoff_ms: 1_000,
        },
    )?;
    Ok(())
}

/// The two-phase scenario: load v1, replay phase 1, block for a real SIGHUP,
/// reload the (now v2) policy from the same path, replay phase 2.
///
/// SIGHUP handling is the systems-programming content here: `signal-hook`
/// installs an async-signal-safe handler that only flips an `AtomicBool` (no
/// allocation, no Lua call, no I/O inside the signal handler itself), and
/// this thread polls that flag instead of doing any real work inside signal
/// context -- the same discipline ch12 uses for signalfd/self-pipe. The flag
/// is registered *before* we announce readiness, so a SIGHUP that lands the
/// instant after the "awaiting" line prints is never missed.
fn cmd_run(policy_path: &str) -> Result<()> {
    // mlua::Error is not Send+Sync in this (non-"send"-feature, single
    // threaded) build, so it cannot implement std::error::Error the way
    // anyhow::Context wants; fold it into an anyhow::Error by its Display
    // text instead of trying to attach `.context()` directly to an
    // `mlua::Result`.
    let lua = build_sandbox()
        .map_err(|e| anyhow::anyhow!("{e}"))
        .context("building sandbox")?;
    load_policy(&lua, policy_path)
        .map_err(|e| anyhow::anyhow!("{e}"))
        .context("loading policy")?;
    phase1(&lua)
        .map_err(|e| anyhow::anyhow!("{e}"))
        .context("phase 1 decisions")?;

    let reload = Arc::new(AtomicBool::new(false));
    signal_hook::flag::register(signal_hook::consts::SIGHUP, reload.clone())
        .context("registering SIGHUP handler")?;

    println!("pmon: awaiting SIGHUP to reload policy");
    while !reload.load(Ordering::Relaxed) {
        thread::sleep(Duration::from_millis(20));
    }
    println!("pmon: reload requested");

    // Rebuild the whole Lua state from scratch on reload, rather than mutate
    // the old one in place: a fresh sandbox guarantees nothing from the old
    // policy (globals, closures) survives into the new one.
    let lua = build_sandbox()
        .map_err(|e| anyhow::anyhow!("{e}"))
        .context("rebuilding sandbox for reload")?;
    load_policy(&lua, policy_path)
        .map_err(|e| anyhow::anyhow!("{e}"))
        .context("reloading policy")?;
    phase2(&lua)
        .map_err(|e| anyhow::anyhow!("{e}"))
        .context("phase 2 decisions")?;

    Ok(())
}

/// Load `policy_path` under the sandbox and run one probe `on_exit` call.
/// Returns the policy's declared version on success; the `mlua::Error` on
/// failure already carries the Lua-side message (a syntax error's location,
/// or -- for `policy-evil.lua` -- "attempt to index a nil value (global
/// 'os')", which is how the sandbox rejection gets reported without the host
/// knowing anything about what the policy tried to do).
fn run_check(policy_path: &str) -> mlua::Result<String> {
    let lua = build_sandbox()?;
    load_policy(&lua, policy_path)?;

    let version: String = lua.globals().get("POLICY_VERSION").unwrap_or_default();

    let probe = ExitEvent {
        name: "probe",
        exit_code: 1,
        restarts: 0,
        consecutive_failures: 0,
        uptime_ms: 9_000,
        last_backoff_ms: 0,
    };
    let info = info_table(&lua, &probe)?;
    let on_exit: Function = lua.globals().get("on_exit")?;
    let _decision: Table = on_exit.call(info)?;

    Ok(version)
}

fn cmd_check(policy_path: &str) -> ! {
    match run_check(policy_path) {
        Ok(version) => {
            println!("pmon: policy ok version={version}");
            exit(0);
        }
        Err(e) => {
            println!("pmon: policy error: {e}");
            exit(1);
        }
    }
}

const USAGE: &str = "usage: app <run|check> --policy <path>";

/// CLI parse failures are the one class of error that never touches Lua or
/// the filesystem, so they get the plain `bail!` treatment instead of the
/// `mlua::Result` plumbing used everywhere else.
fn parse_args(args: &[String]) -> Result<(String, String)> {
    if args.len() < 2 {
        bail!("{USAGE}");
    }
    let sub = args[1].clone();
    if sub != "run" && sub != "check" {
        bail!("{USAGE}");
    }
    let rest = &args[2..];
    if rest.len() != 2 || rest[0] != "--policy" {
        bail!("{USAGE}");
    }
    Ok((sub, rest[1].clone()))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let (sub, policy_path) = match parse_args(&args) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            exit(2);
        }
    };

    match sub.as_str() {
        "run" => {
            if let Err(e) = cmd_run(&policy_path) {
                eprintln!("pmon: error: {e:?}");
                exit(1);
            }
        }
        "check" => cmd_check(&policy_path),
        _ => unreachable!("parse_args only returns run|check"),
    }
}
